### ISSUE TITLE
Trenchbroom support

### ISSUE THREAD

**Issue by jdolan** [jdolan/quetoo#444]
QuakeRanger did a great job bootstrapping Trenchbroom support. There are a few outstanding items before we could comfortably say that we truly support / are supported by Trenchbroom:

- [x] Texture loading, ensure Trenchbroom can support loading multiple formats at the same time
- [x] Entities definition equivalent, ensure it's as accurate and complete as `entities.def`
- [x] Compilation, ensure quemap plays nicely with Trenchbroom
- [x] Model and skin loading, ensure our models look correct

**Comment by kaadmy** [comment]
TrenchBroom currently doesn't even support pk3 or tga files, so before we can even support TrenchBroom it needs patches to work.

**Comment by kaadmy** [comment]
Isn't this mostly related to getting patches in TrenchBroom, not Quetoo itself?

**Comment by jdolan** [comment]
Probably.

**Comment by jdolan** [comment]
It's worth revisiting this. @GibsonGeorge did you mention that you've moved to TB? Would you want to take a peek at current Quetoo support and reassess what we would need to do to provide a good mapping experience?

**Comment by GibsonGeorge** [comment]
Last time I checked this was the state:
- Quetoo.fgd was a perfect port of entities.def
- Textures worked
- Quemap worked but needed to exit when done
- MD3 models didn't have materials
- OBJ models crashed TrenchBroom

This was a while ago and TB's had a few releases since then so some of this might have been addressed, I'll update in a bit.

**Comment by GibsonGeorge** [comment]
Ok, all that's left to go now is model and skin support (MD3s load fine, OBJs load via the obj_neverball type but are scaled wrong, neither have skins)

**Comment by jdolan** [comment]
Thank you for the update on this! It sounds like the best path forward (based on follow-on Discord conversations) is:

- [ ] Scaling issues for obj
- [ ] Include diffuse texture `usemtl` references in our .obj models so that skins work in TrenchBroom (this will fix Radiant too)
- [ ] Convert all static .md3 meshes (e.g. ammo boxes) to .obj so that we don't need to add Q3 shader files for them
- [ ] Ensure we keep Quetoo.fgd in lockstep with entities.def

I will take on the last item there, as I'm in the code all the time, and that's the source of truth for the entity definitions. I already maintain entities.def.

Do we have a volunteer to convert and touch up our models?

**Comment by jdolan** [comment]
@ypsvlq you're currently using TB with Quetoo, right? How are we looking here? Is this ticket still valid, or can we advertise Trenchbroom support at this point?

**Comment by jdolan** [comment]
Just reviewing our Trenchbroom config this morning, I spot a number of issues that should be addressed by someone actively using Trenchbroom:

- [ ] Texture format is only specifying tga, when in fact we also support png, jpeg. Does TB support multiple formats?
- [ ] Texture format is still requesting colormap.pcx. This file is gone and we no longer use a palette.
- [ ] Many entity definitions have stale spawnflags, and lack any of the copy / description found in our Radiant entities.def.
- [ ] Several entity definitions have not yet been ported to Quetoo.fgd, including all client-side `misc_*` entities (fog, dust, sparks, sound, model, ..)

**Comment by jdolan** [comment]
![image](https://user-images.githubusercontent.com/643118/150708046-e4268607-3ce0-4df5-a24a-982f1ed3c253.png)
![quetoo019](https://user-images.githubusercontent.com/643118/150709585-2037b07e-5dcd-4d66-9094-11fa59c5d64f.jpg)

At long last, I'm calling this done! All Quetoo.fgd entities have been brought in-line with their entities.def counterparts. Many improvements have been made, such as specifying the correct types and default values for entity values. The game config itself has been fixed to correctly map to our contents and surface flags, our asset types, our default paths, etc. It's _mostly_ a turn-key process to setup TB for Quetoo now.

As far as viewing entity models in the editor, because many of our .obj meshes are out of scale, it's just not practical right now. The models should all be re-sized to Quake units, but.. that's a rainy day project. On top of that, TrenchBroom has limited .obj support right now, and does not seem to want to load materials from valid .mtl definitions, either. So, for now, all of our entities simply use color-coded bounding boxes.

Lastly, Trenchbroom does not yet support multiple game paths (official vs user-provided). That means that mappers will probably have to target their user-specific directory as the "game path", and then symlink (or copy on Windows, ew) some of the game content into that folder (textures-common.pk3, etc). It kind of sucks, but hopefully TB will address this soon. I've opened an issue here: https://github.com/TrenchBroom/TrenchBroom/issues/3937
